### PR TITLE
[R3C-465] MCH: customize the  completion policy in pedestals workflow

### PR DESCRIPTION
**Pull request by aferrero2707**
This PR fixes the missing end-of-stream messages in the MCH pedestals calibration workflow.

### CODE REVIEW

File: Detectors/MUON/MCH/Calibration/src/MCHChannelCalibratorSpec.h
Comment (aphecetche) [inline Detectors/MUON/MCH/Calibration/src/MCHChannelCalibratorSpec.h:91]: can we get an answer to this question and drop the comment here ?

File: Detectors/MUON/MCH/Calibration/src/channel-calib-workflow.cxx
Comment (aphecetche) [inline Detectors/MUON/MCH/Calibration/src/channel-calib-workflow.cxx]: The name used here must be in sync with what's used in the calibrator spec, right ? If that's so a constant shared by the relevant pieces of code would be less error prone I guess.

File: Detectors/MUON/MCH/Calibration/src/pedestals-decoding-workflow.cxx
Comment (aphecetche) [inline Detectors/MUON/MCH/Calibration/src/pedestals-decoding-workflow.cxx]: do we still need to patch the rdh ? I thought the decoder should now deduce automatically the ul version ? what is missing to drop the patchPage completely ?

File: Detectors/MUON/MCH/Calibration/src/pedestals-decoding-workflow.cxx
Comment (aphecetche) [inline Detectors/MUON/MCH/Calibration/src/pedestals-decoding-workflow.cxx]: same remark as above wrt hard-coded name`calib-mch-pedestal`

File: Detectors/MUON/MCH/Calibration/src/pedestals-decoding-workflow.cxx
Comment (aphecetche) [inline Detectors/MUON/MCH/Calibration/src/pedestals-decoding-workflow.cxx]: same remark as above wrt hard-coded namecalib-mch-pedestal

File: Detectors/MUON/MCH/Calibration/src/MCHChannelCalibratorSpec.h
Comment (shahor02) [reply Detectors/MUON/MCH/Calibration/src/MCHChannelCalibratorSpec.h:91]: We have actually 2 overlapping counters: the one you use from the DataProcessingHeader and `DataHeader::tfCounter``.
I assume the DataProcessingHeader will disappear soon, so would suggest the 2nd one, but better @ktf confirms.

Comment (ktf) [reply Detectors/MUON/MCH/Calibration/src/MCHChannelCalibratorSpec.h:91]: Yes, you probably want to use `DataHeader::tfCounter`, however, as discussed, this is just a counter.

**Comment by aferrero2707** [comment]
@aphecetche I am implementing the changes you suggested, and I am also changing a bit the naming and location of the classes and source files in order to harmonize it with the rest of the MCH code. I will hopefully push an update later today.

File: Detectors/MUON/MCH/Calibration/src/MCHChannelCalibratorSpec.h
Comment (chiarazampolli) [reply Detectors/MUON/MCH/Calibration/src/MCHChannelCalibratorSpec.h:91]: Also in TOF I used this as a proxy for the start time of the TF. Is there a better way to retrieve it? 
See:

https://github.com/AliceO2Group/AliceO2/blob/be56bfe25603983275b909970f2b4128e04eef57/Detectors/TOF/calibration/testWorkflow/TOFChannelCalibratorSpec.h#L72

Here I would really need the start time. Should I use DataProcessingHeader->startTime? I am not sure, also considering that @shahor02 mentioned that it will disappear. 

Then I use it as a counter here:

https://github.com/AliceO2Group/AliceO2/blob/be56bfe25603983275b909970f2b4128e04eef57/Detectors/TOF/calibration/testWorkflow/LHCClockCalibratorSpec.h#L51
https://github.com/AliceO2Group/AliceO2/blob/be56bfe25603983275b909970f2b4128e04eef57/Detectors/TOF/calibration/testWorkflow/TOFCalibCollectorSpec.h#L53

Should we simply replace this call with

`auto tfcounter = o2::header::get<o2::framework::DataHeader*>(pc.inputs().get("input").header)->tfCounter`

? Note also that everybody is "copying" from this (e.g. Detectors/EMCAL/calibration/testWorkflow/EMCALChannelCalibratorSpec.h), so if it should be changed, we need to do it everywhere.

Thanks for any clarification...

Comment (shahor02) [reply Detectors/MUON/MCH/Calibration/src/MCHChannelCalibratorSpec.h:91]: @chiarazampolli strictly speaking, this startTime is the time of the STF building on the FLP machine, which is not the same as the "CTP" time corresponding to the 1st orbit of the TF (which should eventually go to the CCDB and be used for queries).
The real time stamp should be calculated as "ctp_time_of_orbit0" + BCcount_of_TF_1st_orbit, but this is not available at the moment. So, for time-like parameter one can use ``startTime`` until the real one will be available. 
Then can be fixed everywhere in a centralized way.

Comment (chiarazampolli) [reply Detectors/MUON/MCH/Calibration/src/MCHChannelCalibratorSpec.h:91]: Hi @shahor02 , 

Ok, but then I would not change it anywhere, to not make mistakes. Even if for the TimeSlotCalibration a counter can do the same job as a real timestamp, also there we assumed to use eventually use timestamps (note that now that we simulate the input, this startTime is indeed working as a counter, nothing more). What I have doubts about is the fact that you said that 

> DataProcessingHeader will disappear soon

One thing we can do is to change that comment to something like "should be replaced with real start time of the TF".

Chiara

Comment (shahor02) [reply Detectors/MUON/MCH/Calibration/src/MCHChannelCalibratorSpec.h:91]: There were plans to get rid of the DataProcessingHeader completely... DD will not be able to put in any data header the "real start time" of the TF, instead, it can be calculated based on the TF 1st orbit. I think the best would be to calculate it in the workflow, either in the DPL or by dedicated device and feed it as a message.

**Comment by aferrero2707** [comment]
@aphecetche @chiarazampolli @shahor02 would any of you be able to double-check the pedestal data processing using the script(s) provided [here](https://github.com/AliceMCH/AliceO2/tree/mch-fix-pedestals-workflow-completion-policy/Detectors/MUON/MCH/Calibration/test)?
A reference data set is available on the CERNBox: https://cernbox.cern.ch/index.php/s/PuUk2Fu91wqnt4D

I would prefer that someone confirms that the processing is OK before merging. The best would be that you provide me the output of the `test-mch-pedestal-calibration.sh` script so that I can check that everything looks ok.

Thanks!

File: Detectors/MUON/MCH/Calibration/test/run-tfbuilder.sh
Comment (chiarazampolli) [inline Detectors/MUON/MCH/Calibration/test/run-tfbuilder.sh]: Ciao @aferrero2707 , 

You have `TPC`, but it should be `MCH`...

Chiara

Comment (chiarazampolli) [reply Detectors/MUON/MCH/Calibration/test/run-tfbuilder.sh]: and I think you should have:
`transport=$TRANSPORT`
(even though this does not really matter)

**Comment by chiarazampolli** [comment]
Ciao @aferrero2707 , 

Apart from the change `TPC` --> `MCH`, everything seems to work. It prints something like this:
```
[655684:calib-mch-pedestal]: [23:15:56][INFO] Correctly handshaken websocket connection.
[655683:mch-pedestal-decoder]: [23:15:56][INFO] Correctly handshaken websocket connection.
[655682:readout-proxy]: [23:15:57][INFO] readout-proxy[0]: in: 10 (29.3871 MB) out: 0 (0 MB)
[655683:mch-pedestal-decoder]: [23:15:57][INFO] Processed 850102 digits in 11 time frames
[655684:calib-mch-pedestal]: [23:15:57][INFO] received 850102 digits in 11 time frames
[655682:readout-proxy]: [23:15:58][INFO] readout-proxy[0]: in: 1 (2.93876 MB) out: 0 (0 MB)
[655683:mch-pedestal-decoder]: [23:15:58][INFO] Processed 77282 digits in 1 time frames
[655684:calib-mch-pedestal]: [23:15:58][INFO] received 77282 digits in 1 time frames
[655682:readout-proxy]: [23:15:59][INFO] readout-proxy[0]: in: 1 (2.93866 MB) out: 0 (0 MB)
[655682:readout-proxy]: [23:16:00][INFO] readout-proxy[0]: in: 1 (2.93876 MB) out: 0 (0 MB)
[655683:mch-pedestal-decoder]: [23:16:00][INFO] Processed 154564 digits in 2 time frames
[655684:calib-mch-pedestal]: [23:16:00][INFO] received 154564 digits in 2 time frames
[655682:readout-proxy]: [23:16:01][INFO] readout-proxy[0]: in: 1 (2.93876 MB) out: 0 (0 MB)
[655684:calib-mch-pedestal]: [23:16:02][INFO] Checking statistics
[655684:calib-mch-pedestal]: [23:16:02][INFO] Finalize slot 0 <= TF <= 14
[655684:calib-mch-pedestal]: [23:16:02][INFO] S 184  DS 5  CH 1  ENTRIES 195  PED 59.2308  RMS 2.13231
[655684:calib-mch-pedestal]: [23:16:02][INFO] S 184  DS 5  CH 6  ENTRIES 195  PED 102.405  RMS 2.17326
[655684:calib-mch-pedestal]: [23:16:02][INFO] S 426  DS 0  CH 0  ENTRIES 390  PED 81.1821  RMS 5.65324
[655684:calib-mch-pedestal]: [23:16:02][INFO] S 426  DS 0  CH 2  ENTRIES 390  PED 81.4538  RMS 4.60342
[655684:calib-mch-pedestal]: [23:16:02][INFO] S 426  DS 0  CH 3  ENTRIES 390  PED 83.9103  RMS 6.64481
...
```
See:
[out.txt](https://github.com/AliceO2Group/AliceO2/files/6224643/out.txt)

Cheers,
Chiara

**Comment by shahor02** [comment]
@aferrero2707 I had to attach the `o2-calibration-ccdb-populator-workflow  ${ARGS_ALL}` to the workflow in test-mch-pedestal-calibration.sh and limit number of built TFs to 40 in the `run-tfbuilder.sh`. The output is http://ccdb-test.cern.ch:8080/JSRoot?f=http%3A%2F%2Fccdb-test.cern.ch%3A8080%2Fdownload%2F97ac9ab0-90d4-11eb-beb2-200114580202&n=o2%3A%3Amch%3A%3ADsChannelGroup_1617052831505.root

BTW, the `eos:***/INFORMATION` seems to stop terminate correctly the workflow but not the readout-proxy.

**Comment by chiarazampolli** [comment]
Hello, 

I confirm the same observation as Ruben about the `readout-proxy`, I discussed it already with @aferrero2707 , but we did not know if it was intended. Also, I had noticed that to limit the number of TF, one needs the `--control=static` option, or it does not work. Is it expected?

Note also that I had to add `export DATADIST_NO_MLOCK=1` 

I have one more question: In the previous version of the scripts, I was not able to get any output. What did you change? 
I noticed that now in `run-tfbuilder`  you have `address=ipc://@tf-builder-pipe-0` instead of `address=ipc:///tmp/stf-builder-dpl-pipe-0`, is it this? 
You also have added `--readout-proxy` to the configuration of the `o2-dpl-raw-proxy` which was not there before, and which I don't use for the aggregator. Is it mandatory?
Then, I don't see any other difference with respect to when it was not working, at least in the configuration.

Chiara

File: Detectors/MUON/MCH/Calibration/src/pedestals-decoding-workflow.cxx
Comment (aferrero2707) [reply Detectors/MUON/MCH/Calibration/src/pedestals-decoding-workflow.cxx]: @aphecetche you are right, we can now remove the patching of the RDH. I will update the code.

**Comment by aferrero2707** [comment]
@shahor02 thanks for checking. I have tied to access your output file through the link you provided, but the page is empty. Is that normal? Do you see entries in the root file you generated?

For me the other issues observed (for example the readout-proxy not stopping) are independent from the code in this PR, and should be investigated separately.

Any objection to merge this once I have incorporated [this suggestion](https://github.com/AliceO2Group/AliceO2/pull/5772#discussion_r600278892) from @aphecetche and [those](https://github.com/AliceO2Group/AliceO2/pull/5772#pullrequestreview-623630865) from @chiarazampolli?

**Comment by shahor02** [comment]
@aferrero2707 yes, the link take me to JS root session. You can also download it via
http://ccdb-test.cern.ch:8080/download/97ac9ab0-90d4-11eb-beb2-200114580202 or browse the whole dir. on 
http://ccdb-test.cern.ch:8080/browse/MCH/BadChannelCalib?report=true (see 2 objects created yesterday evening).

OK for me to merge this PR after changes

**Comment by chiarazampolli** [comment]
Hi,
Ok for me too to merge after the changes. 

Chiara

**Comment by aphecetche** [comment]
Hi, @aferrero2707, I'm coming late to the party... Tried it and have no new information I guess, but just to restate : 

- had to set `export DATADIST_NO_MLOCK=1` (for the record, tried on an Ubuntu 20.04 virtual machine)
- I plugged the ccdb-populator. To get some output to the CCDB (see in http://ccdb-test.cern.ch:8080/browse/MCH/BadChannelCalib the latest entries with "testLA;coucou" in the metadata :wink:) I either had to limit the number of processed TFs or CTRL-C the dpl proxy (i.e. stop it in both cases)
- CTRL-C the proxy does not stop the readout proxy

**Comment by aferrero2707** [comment]
@chiarazampolli @shahor02 @aphecetche I have updated the code and script, and also added the QC configuration in the tests folder.

I have tried adding QC in the workflow, but I am getting a crash with a `bus error` that does not seem to be related to the MCH-specific QC code:
```
[113437:QC-TASK-RUNNER-QcTaskMCHPreclusters]: [08:55:41][ERROR] Header is not a DataHeader?
[113437:QC-TASK-RUNNER-QcTaskMCHPreclusters]: 
[113437:QC-TASK-RUNNER-QcTaskMCHPreclusters]:  *** Break *** bus error
```

Would you be able to see if you can reproduce the problem?
@Barthelemy any idea of what could cause this crash? I suspect that somewhere a NULL pointer to a DataHeader is being used...

**Comment by Barthelemy** [comment]
@aferrero2707 The error message comes from DPL. The question is where does it crash? is there a stack trace ?

**Comment by aferrero2707** [comment]
@Barthelemy indeed the crash happens deeply in the DPL/FairMQ code. Here is a full stack trace: https://cernbox.cern.ch/index.php/s/sCLUdZxIwrkRr6r

Ping @ktf as well...

**Comment by shahor02** [comment]
@aferrero2707 usually the `[ERROR] Header is not a DataHeader?` appears when some workflows of the piped global workflow have  `--session default` and some don't. Make sure all have the same.

**Comment by aferrero2707** [comment]
@shahor02 absolutely! I was forgetting the `--session default` option for QC, now I have no more crashes.
@Barthelemy @ktf sorry for the noise...

**Review by aphecetche — APPROVED** [review]

**Pull request merged**
